src/pontibus/utils/system_solvation.py

```diff
@@ -375,3 +392,105 @@ def packmol_solvation(
             working_directory=None,
             retain_working_files=False,
         )
+
+
+def openmm_solvation(
+    solute_topology: Topology,
+    solvent_offmol: OFFMolecule,
+    solvation_settings: OpenMMSolvationSettings,
+    neutralize: bool,
+    ion_concentration: Quantity,
+) -> Topology:
+    import openmm
+    import openmm.app
+    from openff.toolkit import Molecule
+
+    ions = [Molecule.from_smiles("[Na+]"), Molecule.from_smiles("[Cl-]")]
+
+    def make_vec3(positions: Quantity) -> openmm.Vec3:
+        return [
+            openmm.Vec3(float(row[0]), float(row[1]), float(row[2]))
+            for row in positions.m_as("nanometer")  # type: ignore[union-attr]
+        ]
+
+    _, n_solvent, box_vectors = _process_inputs(
+        solute_topology,
+        solvent_offmol,
+        solvation_settings,
+        ion_concentration,
+    )
+
+    if not solvent_offmol.is_isomorphic_with(offmol_water):
+        errmsg = "Cannot neutralize a system with non-water solvent"
+        raise ValueError(errmsg)
+
+    if neutralize:
+        if not ion_concentration.is_compatible_with("mole / liter"):
+            errmsg = f"{ion_concentration} is not compatible with mole / liter"
+            raise ValueError(errmsg)
+
+        modeller = openmm.app.Modeller(
+            topology=solute_topology.to_openmm(),
+            positions=make_vec3(solute_topology.get_positions()),
+        )
+
+        from openmmforcefields.generators import SMIRNOFFTemplateGenerator
+
+        forcefield = openmm.app.ForceField("amber14-all.xml", "amber14/tip3pfb.xml")
+
+        forcefield.registerTemplateGenerator(
+            SMIRNOFFTemplateGenerator(molecules=solute_topology.molecule(0)).generator
+        )
+
+        modeller.addSolvent(
+            forcefield=forcefield,
+            model="tip3p",
+            boxVectors=make_vec3(box_vectors),
+            boxShape=solvation_settings.box_shape,
+            positiveIon="Na+",
+            negativeIon="Cl-",
+            ionicStrength=ion_concentration.to_openmm(),  # type: ignore[attr-defined]
+            neutralize=True,
+        )
+
+        topology = Topology.from_openmm(
+            modeller.topology,
+            unique_molecules=[*solute_topology.molecules] + [solvent_offmol] + ions,
+            positions=modeller.getPositions(),
+        )
+
+        solvent_key = solvent_offmol.properties["key"]
+
+        for molecule_index, molecule in enumerate(topology.molecules):
+            if molecule_index < solute_topology.n_molecules:
```

Comment:
Because atom order is preserved and openmm's add_solvent only APPENDS to the atom list, we can identify "solute waters" (that is, assign the correct value to `properties['keys']`) by counting them off in the atom list expecting matching indices.  

(blocking) ensure that failures in this logic are caught by tests. There may be one that counts mols/atoms in different components that would raise the alarm here but should confirm.

Reply:
See comment above & #145 for likely fix.